crates/bitcoind_rpc/src/lib.rs

```diff
@@ -0,0 +1,306 @@
+//! mempool.
+//!
+//! # [`Iter`]
+//!
+//! [`Emitter::into_iterator<B>`] transforms the emitter into an iterator that either returns
+//! [`EmittedBlock`] or [`EmittedHeader`].
+//!
+//! ```rust,no_run
+//! use bdk_bitcoind_rpc::{EmittedBlock, Emitter};
+//! # let client: bdk_bitcoind_rpc::bitcoincore_rpc::Client = todo!();
+//!
+//! for r in Emitter::new(&client, 709_632).into_iterator::<EmittedBlock>() {
+//!     let emitted_block = r.expect("todo: handle error");
+//!     println!(
+//!         "block {}: {}",
+//!         emitted_block.height,
+//!         emitted_block.block.block_hash()
+//!     );
+//! }
+//! ```
+#![warn(missing_docs)]
+
+use std::{collections::BTreeMap, marker::PhantomData};
+
+use bitcoin::{block::Header, Block, BlockHash, Transaction};
+pub use bitcoincore_rpc;
+use bitcoincore_rpc::bitcoincore_rpc_json;
+
+/// Represents a transaction that exists in the mempool.
+pub struct MempoolTx {
+    /// The transaction.
+    pub tx: Transaction,
+    /// Time when transaction first entered the mempool (in epoch seconds).
+    pub time: u64,
+}
+
+/// A block obtained from `bitcoind`.
+pub struct EmittedBlock {
+    /// The actual block.
+    pub block: Block,
+    /// The height of the block.
+    pub height: u32,
+}
+
+/// A block header obtained from `bitcoind`.
+pub struct EmittedHeader {
+    /// The actual block header.
+    pub header: Header,
+    /// The height of the block header.
+    pub height: u32,
+}
+
+/// A structure that emits data sourced from [`bitcoincore_rpc::Client`].
+///
+/// Refer to [module-level documentation] for more.
+///
+/// [module-level documentation]: crate
+pub struct Emitter<'c, C> {
+    client: &'c C,
+    start_height: u32,
+
+    emitted: BTreeMap<u32, BlockHash>,
+    last_block: Option<bitcoincore_rpc_json::GetBlockResult>,
+    /// Records the latest first-seen epoch of emitted mempool transactions.
+    ///
+    /// This allows us to avoid re-emitting some mempool transactions. Mempool transactions need to
+    /// be re-emitted if the latest block that may contain the transaction's ancestors have not yet
+    /// been emitted.
+    last_mempool_time: usize,
+}
+
+impl<'c, C: bitcoincore_rpc::RpcApi> Emitter<'c, C> {
+    /// Constructs a new [`Emitter`] with the provided [`bitcoincore_rpc::Client`].
+    ///
+    /// `start_height` is the block height to start emitting blocks from.
+    pub fn new(client: &'c C, start_height: u32) -> Self {
+        Self {
+            client,
+            start_height,
+            emitted: BTreeMap::new(),
+            last_block: None,
+            last_mempool_time: 0,
+        }
+    }
+
+    /// Emit mempool transactions.
+    ///
+    /// This avoids re-emitting transactions (where viable). We can do this if all blocks
+    /// containing ancestor transactions are already emitted.
+    pub fn mempool(&mut self) -> Result<Vec<MempoolTx>, bitcoincore_rpc::Error> {
```

Comment:
We should have the following tests checking the behavior when `Emitter::mempool` is called multiple times.

- [ ] The receiver (`bdk_chain` structures) is synced to the chain tip, and there is txs in the mempool. When we call `Emitter::mempool` multiple times, mempool txs should not be re-emitted.

- [ ] In the above scenario, if the network's chain extends with empty blocks (and our receiver stays up to sync), we expect that the txs do not get re-emitted.

- [ ] The network is at height `h`. The receiver is synced to height `h-1`. When a mempool tx `A` is introduced at height `h`, we should always re-emit it.

- [ ] In the above scenario, if the receiver syncs up to `h`, we should still re-emit (right after). However, subsequent calls to `Emitter::mempool` should not re-emit.

Reply:
I added a test for the 3rd checkbox and a CI fix in #1140, if it looks OK feel free to copy/paste or cherry-pick, which ever is easier.

Reply:
@notmandatory Thank you for the PR! Unfortunately we duplicated work, and I think my solution covers a broader set of situations, so will stick with that.

Reply:
No problem! closed my PR.